### PR TITLE
[Gameplay] AttackerTactic and Skills 

### PR DESCRIPTION
**Pull request by williamckha**
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    double sequence_length = static_cast<double>(sequence_.size() - 1);
+
+    while (sequence_.size() > 1)
+    {
+        auto [from, to] = sequence_.back();
+        sequence_.pop_back();
+
+        double weight = adj_matrix_[from][to];
+        double adjustment_resistance =
+            sequence_score * weight * weight / ADJUSTMENT_RESISTANCE;
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/scoring/skills/skill_graph.cpp:90]: I think your missing a call to `min` here it's implicit and/or I just can't see it. Comparing to the linked desmos, Curious on how you designed it, looks cool.

**Review by Bvasilchikov — COMMENTED** [review]
devs: *adds one matrix and one loss function*
bazel: "yea so im gonna be 1000 more lines of boilerplate code spread out of over 100 files, have fun!"
devs (I can only imagine...):  (ﾉಥ益ಥ）ﾉ﻿ ┻━┻

For real though, awesome work!! Just 1 comment about inconsistency in the func. Maybe could look into a tool or writing a script that allows us to do these kind of refactoring a bit quicker, even if it is crude, what do you think @itsarune?

**Comment by williamckha** [comment]
> devs: *adds one matrix and one loss function*
> bazel: "yea so im gonna be 2000 more lines of boilerplate code spread out of over 150 files, have fun!"
> devs (I can only imagine...):  (ﾉಥ益ಥ）ﾉ﻿ ┻━┻
> 
> For real though, awesome work!! Just 1 comment about inconsistency in the func. Maybe could look into a tool or writing a script that allows us to do these kind of refactoring a bit quicker, even if it is crude, what do you think @itsarune?

It's actually not all boilerplate lol. This pr has a bunch of gameplay changes so far (including those from #3111) which is why it's this huge mess. I've also been migrating existing tactics (dribble, pivotkick, etc.) to a new skills framework and pushing the changes here

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp (hunk 1 above)
Comment (williamckha) [reply src/software/ai/evaluation/scoring/skills/skill_graph.cpp:90]: I changed the equation a while back but looks like the desmos didn't update, I'll fix that

**Comment by Bvasilchikov** [comment]
absolutely! was just looking at it relatively :). 

I was planning to do another pass through actually because I definitely clicked "view" a few times accidentally and didn't stop, and will check for all specific things your talking about

**Comment by williamckha** [comment]
> absolutely! was just looking at it relatively :). 
> 
> I was planning to do another pass through actually because I definitely clicked "view" a few times accidentally and didn't stop, and will check for all specific things your talking about 

Yeah you don't have to review everything since a lot of the changes are just small and/or temporary solutions to get AI building and running. Once more skills are in place, then we can actually start removing tactics and plays we don't need which should reduce the amount of clutter in the git diff

**Comment by itsarune** [comment]
Why did you remove skill.proto? It was being used in the Generic Factory to come up with new skills.

**Comment by williamckha** [comment]
> Why did you remove skill.proto? It was being used in the Generic Factory to come up with new skills.

I've reworked the Skills, they're more like tactics now. Correct me if im wrong but I dont think the generic factory has anything to do with protobufs. We would need skill protobufs if we want to send them to thunderscope for visualization, haven't got around to doing that yet

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    // The size of the adjacency matrix must be one more than the size of nodes_
+    // since there is a "start" node not explicitly represented by any Skill
+    // that needs to be accounted for
+    adj_matrix_ = std::vector<std::vector<double>>(
+        nodes_.size() + 1, std::vector<double>(nodes_.size() + 1, DEFAULT_EDGE_WEIGHT));
+
+    // The first node of the sequence should always be the "start" node, which has
+    // an ID of nodes_.size()
+    sequence_ = {std::make_pair(nodes_.size(), nodes_.size())};
```
Comment (itsarune) [inline src/software/ai/evaluation/scoring/skills/skill_graph.cpp:27]: how is this assumption enforced?

Comment (itsarune) [reply src/software/ai/evaluation/scoring/skills/skill_graph.cpp:27]: I see so we implicitly assume that the start node is at node.size(), node.size()

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+void SkillGraph::scoreSequence(double sequence_score)
+{
+    LOG_IF(WARNING, sequence_score < -1 || sequence_score > 1)
+        << "Skill sequence being scored with score outside of range [-1, 1]";
+
+    // Equation used to calculate the edge weight adjustment:
+    // https://www.desmos.com/3d/3d2207d9d3
+
+    double adjustment_sigmoid =
+        sigmoid(sequence_score, 0, ADJUSTMENT_SIGMOID_WIDTH) - 0.5;
```
Comment (itsarune) [inline src/software/ai/evaluation/scoring/skills/skill_graph.cpp:79]: i assume that ADJUSTMENT_SIGMOID_WIDTH affects the aggressiveness of our learning?

can we consider making this a dynamic param?

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
+#include "software/ai/evaluation/scoring/skills/skill_graph.h"
+
+#include "software/util/generic_factory/generic_factory.h"
+
+SkillGraph::SkillGraph(std::shared_ptr<Strategy> strategy)
```
Comment (itsarune) [inline src/software/ai/evaluation/scoring/skills/skill_graph.cpp:5]: iiirc, we will only have a finite number of "Skill" instances over the lifetime of the game?

File: src/software/ai/hl/stp/play/ball_placement_play.cpp [hunk 5]
```diff
@@ -15,7 +16,9 @@ BallPlacementPlay::BallPlacementPlay(const TbotsProto::AiConfig &config,
 void BallPlacementPlay::getNextTactics(TacticCoroutine::push_type &yield,
                                        const World &world)
 {
-    auto place_ball_tactic = std::make_shared<DribbleTactic>(ai_config);
+    auto place_ball_tactic = std::make_shared<TypedAssignedSkillTactic<DribbleSkillFSM>>(
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play.cpp]: we should have a macro that automatically creates tactic versions of skills. A version of this: https://github.com/UBC-Thunderbots/Software/blob/5c40f21b067e8f1aaa2794005d43c1aa8bd8444d/src/software/ai/hl/stp/tactic/tactic.h#L15

File: src/software/ai/hl/stp/play/hardware_challenge_plays/BUILD [hunk 6]
```diff
@@ -4,125 +4,127 @@ package(default_visibility = ["//visibility:public"])
...
-    deps = [
-        "//shared:constants",
-        "//software/ai/hl/stp/play",
-        "//software/ai/hl/stp/tactic/move:move_tactic",
-        "//software/logger",
-        "//software/util/generic_factory",
-    ],
-    alwayslink = True,
-)
+# TODO: Update challenge plays to use new Skills architecture
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/hardware_challenge_plays/BUILD]: we can probably delete the challenge plays.

these were a robocup challenge a while ago

File: src/software/ai/hl/stp/play/play.h [hunk 7]
```diff
@@ -115,6 +115,8 @@ class Play
      */
     virtual void updateTactics(const PlayUpdate& play_update);
 
+    inline virtual void initialize(){};
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/play.h]: is this accidental?

**Review by itsarune — APPROVED** [review]
looks good, left some comments

**Comment by williamckha** [comment]
I'm running into deadlock and performance issues with PassStrategy. I've commented out all the multithreading stuff for now so that I can test the gameplay, I'll take a closer look on the weekend

File: src/software/ai/hl/stp/play/hardware_challenge_plays/BUILD (hunk 6 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/hardware_challenge_plays/BUILD]: done

File: src/software/ai/hl/stp/play/play.h (hunk 7 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/play.h]: Good catch, removed

File: src/software/ai/hl/stp/play/ball_placement_play.cpp (hunk 5 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/ball_placement_play.cpp]: Made AssignedSkillTactic easier to use, so hopefully we don't need a macro

https://github.com/UBC-Thunderbots/Software/blob/592041e3279f16f759aae0242d8f48dff865ffd5/src/software/ai/hl/stp/play/ball_placement_play.cpp#L20

File: src/software/ai/evaluation/scoring/skills/skill_graph.cpp (hunk 4 above)
Comment (williamckha) [reply src/software/ai/evaluation/scoring/skills/skill_graph.cpp:5]: Yeah, SkillGraph creates one of each Skill using the GenericFactory. We get a shared ptr to the next Skill to execute from the SkillGraph. We request primitives for a given robot from the Skill; the Skill will keep track of an FSM for that robot and use that FSM to generate primitives for that robot.

**Comment by itsarune** [comment]
Huh I thought I caught these issues with performance. Is there a specific play that shows these issues that I can reproduce on my end?

**Comment by williamckha** [comment]
> Huh I thought I caught these issues with performance. Is there a specific play that shows these issues that I can reproduce on my end?

Try reverting the commenting/changes I made to PassStrategy and change `KeepAwaySkill::getViability` to return something large like 10, then run AI. I think I might have fixed the deadlock issue but performance is pretty bad on my machine. I'm pretty sure it's because we have to acquire a lock on the World everytime we call PassEvaluator and so `PassStrategy::updateWorld` is blocked on the main thread

**Comment by itsarune** [comment]
I checked your branch and I think there are some plays that are constructing Strategy classes inline in the code. I think that causes multiple pass generators to spawn and operate in parallel. Not sure, I'm trying fixing this locally and see if it helps.

**Comment by itsarune** [comment]
check if this branch improves performance on your end: https://github.com/itsarune/Software/tree/strategy_perf_update

**Comment by williamckha** [comment]
Going to merge this into our staging branch and bring the staging branch up to date with master -- i'll open another pr for some minor changes I want to make

**Pull request merged**